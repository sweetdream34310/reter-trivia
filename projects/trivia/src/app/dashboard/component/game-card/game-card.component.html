<div class="innerli">
<div *ngIf="categoryList" class="T-tagging ellipsis" >{{categoryList.join(", ")}}</div>
  <div class="row">
    <ng-container *ngIf="game.gameOptions.playerMode == PlayerMode.Single">
      <div class="col-xs-12 textcenter clearfix">
        <app-user-card [userId]="user?.userId" [type]="userCardType.USER_WEB_IMAGE_NAME_LOCATION_VERTICAL"></app-user-card>
        <hr>
        <div class="dark">Score {{game.stats[user?.userId].score}}</div>
        <div class="div-badge">
          <ng-template *ngIf="game?.gameOptions.isBadgeWithCategory; else showBadge" ngFor let-i="index"  let-badge [ngForOf]="totalBadges" >
            <div class="sprite {{applicationSettings?.badges[game?.stats[user?.userId]?.badge[i]]?.class ? applicationSettings?.badges[game?.stats[user?.userId]?.badge[i]]?.class : 'score0'}}"></div>
          </ng-template>
          <ng-template #showBadge >
            <div [ngClass]="game.stats[user?.userId].score >= 1?'score1':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[user?.userId].score >= 2?'score2':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[user?.userId].score >= 3?'score3':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[user?.userId].score >= 4?'score4':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[user?.userId].score >= 5?'score5':'score0'" class="sprite"></div>
          </ng-template>

        </div>
      </div>
    </ng-container>

    <ng-container *ngIf="game.gameOptions.playerMode == PlayerMode.Opponent && game.playerIds.length == 1">
      <div class="textcenter clearfix">
        <div class="col-xs-6 div-right-border">
          <app-user-card [userId]="user.userId" [type]="userCardType.USER_WEB_IMAGE_NAME_LOCATION_VERTICAL"></app-user-card>
          <span class="vs"> Vs </span>
        </div>
        <div class="col-xs-6">
          <div class="yourimg ">

            <div class="sprite no-opponent"></div>

          </div>
        </div>
      </div>
      <div class="textcenter clearfix">
        <div class="col-xs-6 div-right-border div-score">
          <div class="dark">Score {{game.stats[user.userId].score}}</div>
        </div>
      </div>
      <div class="text-align clearfix">
        <div class="div-width div-right-border float-left">
          <ng-template *ngIf="game?.gameOptions.isBadgeWithCategory; else showBadgeForSinglePlayer" ngFor let-i="index"  let-badge [ngForOf]="totalBadges" >
            <div class="sprite {{applicationSettings?.badges[game?.stats[user?.userId]?.badge[i]]?.class ? applicationSettings?.badges[game?.stats[user?.userId]?.badge[i]]?.class : 'score0'}}"></div>
          </ng-template>
          <ng-template  #showBadgeForSinglePlayer>
            <div [ngClass]="game.stats[user.userId].score >= 1?'score1':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[user.userId].score >= 2?'score2':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[user.userId].score >= 3?'score3':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[user.userId].score >= 4?'score4':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[user.userId].score >= 5?'score5':'score0'" class="sprite"></div>
          </ng-template>

        </div>

        <div class="div-width text-align padding-right float-right">
          <div class="sprite score0"></div>
          <div class="sprite score0"></div>
          <div class="sprite score0"></div>
          <div class="sprite score0"></div>
          <div class="sprite score0"></div>
        </div>
      </div>
    </ng-container>

    <ng-container *ngIf="game.gameOptions.playerMode == PlayerMode.Opponent && game.playerIds.length > 1">
      <div class="textcenter clearfix">
        <div class="col-xs-6 div-right-border">
          <app-user-card [userId]="user.userId" [type]="userCardType.USER_WEB_IMAGE_NAME_LOCATION_VERTICAL"></app-user-card>
          <span class="vs"> Vs </span>
        </div>
        <div class="col-xs-6">
          <app-user-card [userId]="otherUserId" [type]="userCardType.USER_WEB_IMAGE_NAME_LOCATION_VERTICAL"></app-user-card>
        </div>
      </div>
      <div class="clearfix">
        <div class="col-xs-6 div-right-border div-score score-text-align">
          <div class="dark">Score {{game.stats[user.userId].score}}</div>
        </div>
        <div class="col-xs-6 div-score score-text-align">
          <div class="dark">Score {{game.stats[otherUserId].score}}</div>
        </div>
      </div>
      <div class="text-align clearfix">
        <div class="div-width div-right-border float-left">
          <ng-template  *ngIf="game?.gameOptions?.isBadgeWithCategory; else showBadgeForTwoPlayer" ngFor let-i="index"  let-badge [ngForOf]="totalBadges" >
              <div class="sprite {{applicationSettings?.badges[game?.stats[user?.userId]?.badge[i]]?.class ? applicationSettings?.badges[game?.stats[user?.userId]?.badge[i]]?.class : 'score0'}}"></div>
          </ng-template>
          <ng-template #showBadgeForTwoPlayer >
              <div [ngClass]="game.stats[user.userId].score >= 1?'score1':'score0'" class="sprite"></div>
              <div [ngClass]="game.stats[user.userId].score >= 2?'score2':'score0'" class="sprite"></div>
              <div [ngClass]="game.stats[user.userId].score >= 3?'score3':'score0'" class="sprite"></div>
              <div [ngClass]="game.stats[user.userId].score >= 4?'score4':'score0'" class="sprite"></div>
              <div [ngClass]="game.stats[user.userId].score >= 5?'score5':'score0'" class="sprite"></div>
          </ng-template>

        </div>

        <div class="div-width text-align padding-right float-right">
          <ng-template *ngIf="game?.gameOptions?.isBadgeWithCategory; else showBadgeForTwoPlayerOpponent" ngFor let-i="index"  let-badge [ngForOf]="totalBadges" >
              <div class="sprite {{applicationSettings?.badges[game?.stats[otherUserId].badge[i]]?.class ? applicationSettings?.badges[game?.stats[otherUserId].badge[i]]?.class : 'score0'}}"></div>
          </ng-template>
          <ng-template #showBadgeForTwoPlayerOpponent >
            <div [ngClass]="game.stats[otherUserId].score >= 1?'score1':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[otherUserId].score >= 2?'score2':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[otherUserId].score >= 3?'score3':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[otherUserId].score >= 4?'score4':'score0'" class="sprite"></div>
            <div [ngClass]="game.stats[otherUserId].score >= 5?'score5':'score0'" class="sprite"></div>
          </ng-template>
        </div>
      </div>
    </ng-container>
    <div class="col-xs-12 div-spacing textcenter" *ngIf="game.gameOptions.playerMode != PlayerMode.Single">
      <span class="light">Round {{game.round}} / {{totalRound}}</span>
    </div>
  </div>

  <div class="textcenter">
    <a *ngIf="myTurn && !isHidePlayNow" [routerLink]="['/game-play', game.gameId]" class="playnow btn">
      <span>Your Turn</span>
      <p>Play Now</p>
    </a>
  </div>

  <div *ngIf="myTurn" [class.m-t-44]="isHidePlayNow" class="opponent">
    <p class="dark">{{remainingHours }}:{{remainingMinutes}} hrs</p>
    <p class="ellipsis">remaining to play your turn</p>
  </div>

  <div *ngIf="game.GameStatus === gameStatus.AVAILABLE_FOR_OPPONENT || game.GameStatus === gameStatus.WAITING_FOR_FRIEND_INVITATION_ACCEPTANCE || game.GameStatus === gameStatus.WAITING_FOR_RANDOM_PLAYER_INVITATION_ACCEPTANCE"
    class="opponent">
    <a class="btn-share waiting-for-opponent">Waiting for
      <p>Opponent</p>
    </a>
  </div>

  <div *ngIf="!myTurn && game.GameStatus === gameStatus.WAITING_FOR_NEXT_Q" class="opponent">
    <a class="btn-share opponent-turn">Opponent's Turn </a>
  </div>
</div>