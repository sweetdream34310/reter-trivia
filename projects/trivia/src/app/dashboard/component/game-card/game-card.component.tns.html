<StackLayout class="active-container">
 
    <GridLayout rows="36" columns="*,auto,auto">
        <Label *ngIf="categoryList" text="{{categoryList.join(', ')}}"
            row="0" col="0" class="cat-title"></Label>
        <StackLayout row="0" col="1" horizontalAlignment="right">
            <AbsoluteLayout *ngIf="game?.gameOptions.isBadgeWithCategory">
                <Image *ngFor="let badge of earnedBadgesByOtherUser; let i = index"  left="{{24 + ((earnedBadgesByOtherUser.length - (i+1)) * 12)}}" top="5" src="res://{{applicationSettings?.badges[badge]?.image?.name}}"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image> 
                <StackLayout left="5" top="5">
                    <app-user-card [userId]="otherUserId" [type]="userCardType.USER_IMAGE_30X30" [width]="30" [height]="30">
                    </app-user-card>
                </StackLayout>   
            </AbsoluteLayout>
            <AbsoluteLayout *ngIf="!game?.gameOptions.isBadgeWithCategory">
                <Image *ngIf="game.stats[otherUserId]?.score >= 5" left="74" top="5" src="res://badge5"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>

                <Image *ngIf="game.stats[otherUserId]?.score >= 4" left="62" top="5" src="res://badge4"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>

                <Image *ngIf="game.stats[otherUserId]?.score >= 3" left="50" top="5" src="res://badge3"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>

                <Image *ngIf="game.stats[otherUserId]?.score >= 2" left="38" top="5" src="res://badge2"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>

                <Image *ngIf="game.stats[otherUserId]?.score >= 1" left="25" top="5" src="res://badge1"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>
                <StackLayout left="5" top="5">
                    <app-user-card [userId]="otherUserId" [type]="userCardType.USER_IMAGE_30X30" [width]="30" [height]="30">
                    </app-user-card>
                </StackLayout>
            </AbsoluteLayout>
        </StackLayout>

        <StackLayout row="0" col="2" horizontalAlignment="right">
            <AbsoluteLayout *ngIf="game?.gameOptions.isBadgeWithCategory">
                <Image *ngFor="let badge of earnedBadges; let i = index" left="{{24 + ((earnedBadges.length - (i+1)) * 12)}}" top="5" src="res://{{applicationSettings?.badges[badge]?.image?.name}}"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>
                <StackLayout left="5" top="5">
                    <app-user-card [userId]="user.userId" [type]="userCardType.USER_IMAGE_30X30" [width]="30" [height]="30">
                    </app-user-card>
                </StackLayout>
            </AbsoluteLayout>
            <AbsoluteLayout *ngIf="!game?.gameOptions.isBadgeWithCategory">
                <Image *ngIf="game.stats[user.userId]?.score >= 5" left="74" top="5" src="res://badge5"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>

                <Image *ngIf="game.stats[user.userId]?.score >= 4" left="62" top="5" src="res://badge4"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>

                <Image *ngIf="game.stats[user.userId]?.score >= 3" left="50" top="5" src="res://badge3"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>

                <Image *ngIf="game.stats[user.userId]?.score >= 2" left="38" top="5" src="res://badge2"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>

                <Image *ngIf="game.stats[user.userId]?.score >= 1" left="25" top="5" src="res://badge1"
                    height="30" width="30" marginLeft="2" verticalAlignment="center" marginLeft="2" borderRadius="50"></Image>

                <StackLayout left="5" top="5">
                    <app-user-card [userId]="user.userId" [type]="userCardType.USER_IMAGE_30X30" [width]="30" [height]="30">
                    </app-user-card>
                </StackLayout>
            </AbsoluteLayout>
        </StackLayout>

    </GridLayout>

    <FlexboxLayout horizontalAlignment="left" justifyContent="space-between" width="85%" *ngIf="(game.gameOptions.playerMode == PlayerMode.Opponent && game.playerIds.length > 1) ||
            (game.gameOptions.playerMode == PlayerMode.Single)">
        
                <StackLayout alignSelf="center" orientation="horizontal" [visibility]="myTurn ? 'visible' : 'hidden' ">
                    <Label class="fa active-game-timer" text="&#xf017; {{remainingHours }}:{{remainingMinutes}}" textWrap="true"
                        android:paddingTop=""></Label>
                </StackLayout>
        
                <FlexboxLayout alignSelf="center">
                    <Image src="~/assets/images/mobile/tally.png" width="20" marginRight="2"></Image>
                    <Label text="{{game.round}} / {{totalRound}}" class="active-game-timer" textWrap="true"></Label>
                </FlexboxLayout>
        
                <StackLayout alignSelf="center" orientation="horizontal">
                    <app-user-card *ngIf="myTurn" [userId]="user.userId" [type]="userCardType.USER_IMAGE_30X30" [width]="20"
                        [height]="20">
                    </app-user-card>
                    <app-user-card *ngIf="!myTurn && otherUserId" [userId]="otherUserId" [type]="userCardType.USER_IMAGE_30X30"
                        [width]="20" [height]="20">
                    </app-user-card>
                    <Label *ngIf="myTurn" text="Your Turn " class="active-game-timer"></Label>
                    <Label *ngIf="!myTurn && otherUserId" text="{{otherUserInfo?.displayName}} 's Turn"
                        class="active-game-timer"></Label>
                </StackLayout>
            </FlexboxLayout>

    <Label *ngIf="myTurn && !isHidePlayNow" text="Play Now" textWrap="true" class="active-play-now" [nsRouterLink]="['/game-play', game.gameId]"
        clearHistory="true"></Label>

    <Label *ngIf="(game.GameStatus === gameStatus.AVAILABLE_FOR_OPPONENT || game.GameStatus === gameStatus.WAITING_FOR_FRIEND_INVITATION_ACCEPTANCE || game.GameStatus === gameStatus.WAITING_FOR_RANDOM_PLAYER_INVITATION_ACCEPTANCE)"
        text="Waiting for Opponent" textWrap="true" class="active-waiting-opponent"></Label>

    <Label *ngIf="!myTurn && game.GameStatus === gameStatus.WAITING_FOR_NEXT_Q"
    text="Opponent's Turn" textWrap="true" class="active-waiting-opponent"></Label>
</StackLayout>