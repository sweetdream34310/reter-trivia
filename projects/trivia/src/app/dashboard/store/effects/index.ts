import { DashboardEffects } from './dashboard.effects';

export * from './dashboard.effects';

export const effects = [
  DashboardEffects
];

