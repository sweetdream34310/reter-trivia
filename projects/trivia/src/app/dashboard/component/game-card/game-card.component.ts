import { Component, Input, OnInit, OnChanges, OnDestroy, ChangeDetectorRef, ChangeDetectionStrategy, SimpleChanges } from '@angular/core';
import { Observable, Subscription, timer } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { User, Game, Category, PlayerMode, GameStatus, CalenderConstants, userCardType,
   ApplicationSettings } from 'shared-library/shared/model';
import { Utils } from 'shared-library/core/services';
import { AppState, appState, categoryDictionary } from '../../../store';
import { take } from 'rxjs/operators';
import { AutoUnsubscribe } from 'shared-library/shared/decorators';

@Component({
  selector: 'game-card',
  templateUrl: './game-card.component.html',
  styleUrls: ['./game-card.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})

@AutoUnsubscribe({ 'arrayName': 'subscriptions' })
export class GameCardComponent implements OnInit, OnChanges, OnDestroy {

  @Input() game: Game;
  @Input() cardType: any;
  @Input() categoryDict: { [key: number]: Category };
  @Input() userDict: { [key: string]: User };
  @Input() applicationSettings: ApplicationSettings;
  totalBadges: string[];
  earnedBadges: string[];
  earnedBadgesByOtherUser: string[];
  user$: Observable<User>;
  correctAnswerCount: number;
  questionIndex: number;
  user: User;
  myTurn: boolean;
  otherUserId: string;
  otherUserInfo: User;
  public remainingHours: string;
  public remainingMinutes: string;
  timerSub: Subscription;
  randomCategoryId = 0;
  PlayerMode = PlayerMode;
  totalRound = 16;
  gameStatus: any;
  defaultAvatar = 'assets/images/default-avatar-small.png';
  userDict$: Observable<{ [key: string]: User }>;
  subscriptions = [];
  userCardType = userCardType;
  categoryList = [];
  isHidePlayNow = false;
  constructor(public store: Store<AppState>, public utils: Utils, private cd: ChangeDetectorRef) {
    this.gameStatus = GameStatus;
    this.user$ = this.store.select(appState.coreState).pipe(select(s => s.user));
    this.subscriptions.push(this.user$.subscribe(user => {
      if (user !== null) {
        this.user = user;
      }
      this.cd.markForCheck();
    }));

    this.userDict$ = store.select(appState.coreState).pipe(select(s => s.userDict));
    this.subscriptions.push(this.userDict$.subscribe(userDict => {
      this.userDict = userDict;
      if (this.game && this.user && this.userDict) {
        this.otherUserId = this.game.playerIds.filter(userId => userId !== this.user.userId)[0];
        this.otherUserInfo = this.userDict[this.otherUserId];
      }
      this.cd.markForCheck();
    }));


  }

  ngOnInit() {

    this.totalRound = (Number(this.game.gameOptions.playerMode) === PlayerMode.Single) ? 8 : 16;
    this.subscriptions.push(this.store.select(appState.coreState).pipe(take(1)).subscribe(s => {
      if (s.user && this.user) {
        this.user = s.user;
        this.myTurn = this.game.nextTurnPlayerId === this.user.userId &&
        ([this.gameStatus.STARTED, this.gameStatus.RESTARTED, this.gameStatus.JOINED_GAME, this.gameStatus.WAITING_FOR_NEXT_Q].indexOf(this.game.GameStatus) >= 0 );
        this.randomCategoryId = Math.floor(Math.random() * this.game.gameOptions.categoryIds.length);
        if (this.myTurn) {
          this.updateRemainingTime();
        }
        this.cd.markForCheck();
      }
    }));
  }

  ngOnChanges(changes: SimpleChanges) {
    this.questionIndex = this.game.playerQnAs.length;
    this.correctAnswerCount = this.game.playerQnAs.filter((p) => p.answerCorrect).length;
    if (this.game) {
      this.otherUserId = this.game.playerIds.filter(userId => userId !== this.user.userId)[0];
      this.otherUserInfo = this.userDict[this.otherUserId];
    }
    if (changes.game && changes.game.currentValue && this.game.gameOptions.isBadgeWithCategory) {
      if (this.user.userId && this.game.stats[this.user.userId].badge) {
        this.earnedBadges = [...this.game.stats[this.user.userId].badge].reverse();
      }
      if (Number(this.game.gameOptions.playerMode) === PlayerMode.Opponent && this.game.stats[this.otherUserId] &&
         this.game.stats[this.otherUserId].badge) {
        this.earnedBadgesByOtherUser = [...this.game.stats[this.otherUserId].badge].reverse();
      }
    }
    if (changes.applicationSettings && changes.applicationSettings.currentValue) {
      this.totalBadges =  Object.keys(this.applicationSettings.badges);
    }
    if (this.game && this.categoryDict) {
      this.categoryList = [
        ...this.game.gameOptions.categoryIds
          .map(id =>
            this.categoryDict[id] ? this.capitalizeFirstLetter(this.categoryDict[id].categoryName) : ""
          )
          .filter(name => name !== '')
      ];
    }
  }



  ngOnDestroy() {

  }

  updateRemainingTime() {
    this.timerSub = timer(1000, 1000).subscribe(t => {
      if (this.game.nextTurnPlayerId === this.user.userId) {
        const diff = this.utils.getTimeDifference(this.game.turnAt);
        const hour = Math.floor(diff / (CalenderConstants.HOURS_CALCULATIONS));
        const minute = Math.floor(diff % (CalenderConstants.HOURS_CALCULATIONS) / (CalenderConstants.MINUTE_CALCULATIONS));
        if (hour < 32 ) {
          if (minute > 0) {
            this.remainingHours = this.utils.convertIntoDoubleDigit(31 - hour);
            this.remainingMinutes = this.utils.convertIntoDoubleDigit(60 - minute);
          } else {
            this.remainingHours = this.utils.convertIntoDoubleDigit(32 - hour);
            this.remainingMinutes = this.utils.convertIntoDoubleDigit(0);
          }
        } else {
          this.remainingHours = this.utils.convertIntoDoubleDigit(0);
          this.remainingMinutes = this.utils.convertIntoDoubleDigit(0);
          this.isHidePlayNow = true;
        }
      }
      this.cd.markForCheck();
    });
    this.subscriptions.push(this.timerSub);
  }

  capitalizeFirstLetter(val: string) {
    return val.charAt(0).toUpperCase() + val.slice(1);
  }
}
