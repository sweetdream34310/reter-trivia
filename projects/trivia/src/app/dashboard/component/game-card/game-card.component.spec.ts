import { async, ComponentFixture, TestBed, tick, fakeAsync } from '@angular/core/testing';
import { GameCardComponent } from './game-card.component';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { Store, MemoizedSelector } from '@ngrx/store';
import { provideMockStore, MockStore } from '@ngrx/store/testing';
import { Utils } from 'shared-library/core/services';
import { User, Game, PlayerMode, GameStatus } from 'shared-library/shared/model';
import { AppState, appState } from '../../../store';
import { testData } from 'test/data';
import { CoreState } from 'shared-library/core/store';
import { MatSnackBarModule } from '@angular/material';

describe('GameCardComponent', () => {

    let component: GameCardComponent;
    let fixture: ComponentFixture<GameCardComponent>;
    let spy: any;
    let user: User;
    let mockStore: MockStore<AppState>;

    beforeEach(async(() => {
        TestBed.configureTestingModule({
            declarations: [GameCardComponent],
            schemas: [NO_ERRORS_SCHEMA],
            providers: [
                {provide: Utils, useValue: {
                    getTimeDifference(turnAt: number) {
                        return 1588313130838 - turnAt;
                    },
                    convertIntoDoubleDigit(digit: Number) {
                        return (digit < 10) ? `0${digit}` : `${digit}`;
                    }
                }},
                provideMockStore( {
                    selectors: [
                      {
                        selector: appState.coreState,
                        value: {}
                      }
                    ]
                  })
            ],
            imports: [ MatSnackBarModule ]
        });

    }));

    beforeEach(() => {
        // create component
        fixture = TestBed.createComponent(GameCardComponent);
        // mock data
        mockStore = TestBed.get(Store);
        spy = spyOn(mockStore, 'dispatch');

        component = fixture.debugElement.componentInstance;

        const dbModel = testData.games[0];
        component.game = Game.getViewModel(dbModel);

    });

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it(' updateRemainingTime function should be called from ngOnInit when it is user\'s turn', () => {
        component.updateRemainingTime = jest.fn();
        component.isHidePlayNow = false;
        component.applicationSettings = testData.applicationSettings;

        component.categoryDict = testData.categoryDictionary;
        user = { ...testData.userList[0] };
        mockStore.overrideSelector<AppState, Partial<CoreState>>(appState.coreState, {
            user: user
          });
          mockStore.refreshState();
        component.PlayerMode = PlayerMode;
        component.gameStatus = GameStatus;
        const otherUser = { ...testData.userList[1] };
        component.otherUserId = 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1';
        component.userDict = {'4kFa6HRvP5OhvYXsH9mEsRrXj4o2': user, 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1': otherUser};
        fixture.detectChanges();
        expect(component.updateRemainingTime).toHaveBeenCalledTimes(1);
    });

    it(' capitalizeFirstLetter should be called for all categoryId data from game options', () => {

        component.capitalizeFirstLetter = jest.fn();
        component.applicationSettings = testData.applicationSettings;
        component.categoryDict = testData.categoryDictionary;
        component.user = { ...testData.userList[0] };
        component.PlayerMode = PlayerMode;
        component.gameStatus = GameStatus;
        const otherUser = { ...testData.userList[1] };
        component.otherUserId = 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1';
        component.userDict = {'4kFa6HRvP5OhvYXsH9mEsRrXj4o2': user, 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1': otherUser};
        component.ngOnChanges({
            applicationSettings :
            {
                previousValue: undefined,
                currentValue: testData.applicationSettings,
                firstChange: true,
                isFirstChange: undefined
            }
        });
        expect(component.capitalizeFirstLetter).toHaveBeenCalledTimes(component.game.gameOptions.categoryIds.length);
    });

    it('verify that Other Player information should be set with game data', () => {
        component.applicationSettings = testData.applicationSettings;
        component.user = { ...testData.userList[0] };
        const otherUser = { ...testData.userList[1] };
        component.userDict = {'4kFa6HRvP5OhvYXsH9mEsRrXj4o2': user, 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1': otherUser};
        component.ngOnChanges({
            applicationSettings :
            {
                previousValue: undefined,
                currentValue: testData.applicationSettings,
                firstChange: true,
                isFirstChange: undefined
            }
        });
        expect(component.otherUserId).toEqual(otherUser.userId);
        expect(component.otherUserInfo).toEqual(otherUser);
    });

    it('Verify that Earned badges should be set with Game data', () => {
        component.applicationSettings = testData.applicationSettings;
        component.user = { ...testData.userList[0] };
        const otherUser = { ...testData.userList[1] };
        component.userDict = {'4kFa6HRvP5OhvYXsH9mEsRrXj4o2': user, 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1': otherUser};
        const dbModel = testData.games[1];
        component.game = Game.getViewModel(dbModel);
        const game = Game.getViewModel(dbModel);
        const currentUserEarnedBadges = [...game.stats[component.user.userId].badge].reverse();
        component.PlayerMode = PlayerMode;
        const otherUserEarnedBadges = [...game.stats[otherUser.userId].badge].reverse();
        component.ngOnChanges({
            game :
            {
                previousValue: undefined,
                currentValue: Game.getViewModel(dbModel),
                firstChange: true,
                isFirstChange: undefined
            }
        });
        expect(component.earnedBadges).toEqual(currentUserEarnedBadges);
        expect(component.earnedBadgesByOtherUser).toEqual(otherUserEarnedBadges);
    });

    it('should match snapshot', () => {
        component.isHidePlayNow = false;
        component.applicationSettings = testData.applicationSettings;

        component.categoryDict = testData.categoryDictionary;
        component.user = testData.userList[0];
        component.PlayerMode = PlayerMode;
        component.gameStatus = GameStatus;
        const otherUser = { ...testData.userList[1] };
        component.otherUserId = 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1';
        component.userDict = {'4kFa6HRvP5OhvYXsH9mEsRrXj4o2': user, 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1': otherUser};
        component.ngOnInit();
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();
    });

    it('capitalizeFirstLetter function should capitalize first letter', () => {
        const categoryName = 'infrastructure/networking';
        const categoryNameWithCapitalizedFirstLetter = component.capitalizeFirstLetter(categoryName);
        expect(categoryNameWithCapitalizedFirstLetter).toEqual('Infrastructure/networking');
    });

    it(`User should be undefined when component is created`, () => {
        expect(component.user).toBe(undefined);
    });

    it('User should be set when user is logged in', () => {
        user = { ...testData.userList[0] };
        mockStore.overrideSelector<AppState, Partial<CoreState>>(appState.coreState, {
            user: user
          });
          mockStore.refreshState();
          expect(component.user).toBe(user);
      });

      it('User dictionary should be set when values are emitted', () => {
        user = { ...testData.userList[0] };
        const otherUser = { ...testData.userList[1] };
        const userDict = {'4kFa6HRvP5OhvYXsH9mEsRrXj4o2': user, 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1': otherUser};
        mockStore.overrideSelector<AppState, Partial<CoreState>>(appState.coreState, {
            userDict: userDict
          });
          mockStore.refreshState();
          expect(component.userDict).toBe(userDict);
      });


    it('remaining time should be 2 hr 30 min', (async () => {

        component.isHidePlayNow = false;
        component.applicationSettings = testData.applicationSettings;

        component.categoryDict = testData.categoryDictionary;
        component.user = testData.userList[0];
        component.PlayerMode = PlayerMode;
        component.gameStatus = GameStatus;
        const otherUser = { ...testData.userList[1] };
        component.otherUserId = 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1';
        component.userDict = {'4kFa6HRvP5OhvYXsH9mEsRrXj4o2': user, 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1': otherUser};
        component.updateRemainingTime();

        await new Promise((r) => setTimeout(r, 2000));
        expect(component.remainingMinutes).toBe('30');
        expect(component.remainingHours).toBe('02');
    }));

    it('remaining time should be 0 hr 0 min', (async () => {

        component.isHidePlayNow = false;
        component.applicationSettings = testData.applicationSettings;
        const dbModel = testData.games[1];
        component.game = Game.getViewModel(dbModel);
        component.categoryDict = testData.categoryDictionary;
        component.user = testData.userList[0];
        component.PlayerMode = PlayerMode;
        component.gameStatus = GameStatus;
        const otherUser = { ...testData.userList[1] };
        component.otherUserId = 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1';
        component.userDict = {'4kFa6HRvP5OhvYXsH9mEsRrXj4o2': user, 'yP7sLu5TmYRUO9YT4tWrYLAqxSz1': otherUser};
        component.updateRemainingTime();

        await new Promise((r) => setTimeout(r, 2000));
        expect(component.remainingMinutes).toBe('00');
        expect(component.remainingHours).toBe('00');
    }));

    afterEach(() => {
        fixture.destroy();
    });

});
