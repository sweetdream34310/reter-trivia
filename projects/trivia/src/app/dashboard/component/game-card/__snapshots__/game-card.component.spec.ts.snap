// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameCardComponent should match snapshot 1`] = `
<game-card
  PlayerMode={[Function Object]}
  applicationSettings={[Function Object]}
  categoryDict={[Function Object]}
  categoryList={[Function Array]}
  cd={[Function ViewRef_]}
  defaultAvatar={[Function String]}
  game={[Function Game]}
  gameStatus={[Function Object]}
  isHidePlayNow="false"
  otherUserId={[Function String]}
  randomCategoryId="0"
  store={[Function MockStore]}
  subscriptions={[Function Array]}
  totalRound={[Function Number]}
  user={[Function Object]}
  user$={[Function Store]}
  userCardType={[Function Object]}
  userDict={[Function Object]}
  userDict$={[Function Store]}
  utils={[Function Object]}
>
  <div
    class="innerli"
  >
    
    <div
      class="T-tagging ellipsis"
    >
      
    </div>
    <div
      class="row"
    >
      
      
      
      
      <div
        class="textcenter clearfix"
      >
        <div
          class="col-xs-6 div-right-border"
        >
          <app-user-card />
          <span
            class="vs"
          >
             Vs 
          </span>
        </div>
        <div
          class="col-xs-6"
        >
          <app-user-card />
        </div>
      </div>
      <div
        class="clearfix"
      >
        <div
          class="col-xs-6 div-right-border div-score score-text-align"
        >
          <div
            class="dark"
          >
            Score 10
          </div>
        </div>
        <div
          class="col-xs-6 div-score score-text-align"
        >
          <div
            class="dark"
          >
            Score 1
          </div>
        </div>
      </div>
      <div
        class="text-align clearfix"
      >
        <div
          class="div-width div-right-border float-left"
        >
          
          
          
          
        </div>
        <div
          class="div-width text-align padding-right float-right"
        >
          
          
          
          
        </div>
      </div>
      
      <div
        class="col-xs-12 div-spacing textcenter"
      >
        <span
          class="light"
        >
          Round 3 / 16
        </span>
      </div>
    </div>
    <div
      class="textcenter"
    >
      
    </div>
    
    
    
    <div
      class="opponent"
    >
      <a
        class="btn-share opponent-turn"
      >
        Opponent's Turn 
      </a>
    </div>
  </div>
</game-card>
`;
